File: teamcaspers_scripts/Trainings/HPC_Python/code/Mandelbrot/c/mandelbrot.c
#include <complex.h>
#include <stdio.h>
int init(double complex* q, double xmin, double xmax, double ymin, double ymax, int xres, int yres){
	double dx = (double) (xmax - xmin) / (xres - 1);
	double dy = (double) (ymax - ymin) / (yres - 1);
	int idx = 0;
	for (int i = 0; i < xres; ++i){
		for (int j = 0; j < yres; ++j){
			q[idx++] = xmin + i * dx + (ymin + j * dy) * I;
		}
	}
	return xres * yres;
}


void calculate(int* iterations, double complex* q, int n, int maxiter){
	double complex z;
	double complex c;
	for (int i = 0; i < n; ++i){
		z = 0.0;
		c = q[i];
		for (int j = 0; j < maxiter; ++j){
		       z = z * z + c;
		       if (creal(z * conj(z)) > 4.0){
			       iterations[i] = j;
			       break;
		       }
		}
	}
}

File: teamcaspers_scripts/Trainings/HPC_Python/code/Mandelbrot/fortran/mandelbrot.f90
module mandelbrot

complex, allocatable, dimension(:) :: q
integer, allocatable, dimension(:) :: iterations
integer maximumNumberOfIterations, numberOfPoints

contains
    subroutine init(minRealValue, maxRealValue, minImagValue, maxImagValue, xres, yres, maxiter)

        real, intent(in) :: minRealValue, maxRealValue, minImagValue, maxImagValue
        real :: dx, dy
        integer, intent(in) :: xres, yres, maxiter
        integer ::  idx, i, j
        complex :: XJ = (0, 1)
        
        numberOfPoints = xres * yres
        dx = (maxRealValue - minRealValue) / (xres - 1)
        dy = (maxImagValue - minImagValue) / (yres - 1)
        maximumNumberOfIterations = maxiter
        
        allocate(q(numberOfPoints))
        allocate(iterations(numberOfPoints))
        
        idx = 1
        do i = 0, xres - 1
            do j = 0, yres - 1
                q(idx) = cmplx(minRealValue + i * dx,  minImagValue  + j * dy)
                idx = idx + 1
            end do
        end do
        
    end subroutine init
        
    subroutine calculate()

        complex :: z, c
        integer i, j
        do i = 1, numberOfPoints
            z = (0.0, 0.0)
            c = q(i)
            do j = 1, maximumNumberOfIterations
                z = z * z + c
                if (real(z * conjg(z)) > 4.0) then
                    iterations(i) = j
                    exit
                end if
            end do
        end do
    end subroutine calculate

end module mandelbrot
